### ISSUE TITLE
Using SHAP for feature selection with noisy high-dimensional data

### ISSUE THREAD

**Issue by brunofacca** [shap/shap#1338]
First, thank you very much for this great library. It has been extremely useful to me.

I have a dataset with 5,000 features and multiple (imbalanced) classes. Most of those features are mostly noise and even the ones that have predictive power also have a fair amount of noise on them. The predictive power of some features is affected by interactions between them. Feeding all those features to a classifier such as XGBoost or LightGBM with proper hyperparameters results in predictions that are often better than a random guess but only by a small margin. 

As a feature selection strategy, I'm using SHAP to calculate feature importance scores with XGBoost, LightGBM, and CatBoost and selecting the features with the highest importance. Are those feature importance scores reliable even if the classifiers that generated them have low accuracy (e.g., F1 score on cross-validation)? 

Thank you :bow:

**Comment by HughChen** [comment]
> 1. Are those feature importance scores reliable even if the classifiers that generated them have low accuracy (e.g., F1 score on cross-validation)?

In general, the feature importance scores are only as good as the model they are obtained from.  If the model is totally random, the feature importances will be accurate in the sense that they will explain what features change the model's output.  Unfortunately, if the model only learned spurious relationships, the importance scores will reflect that.

> 2. Instead of just calculating the mean of the 3 classifier's feature importance scores, would it make more sense to give higher weight to the classifiers whose cross-validation F1 scores were higher?
> 3. This is probably out of scope (please feel free to ignore), but would you recommend another feature selection strategy altogether?

One alternative for feature selection is to use the Shapley value of the loss, rather than the model output (I'm not sure if this is supported for multi-class tree objectives though).  There is an example of this in Figure 9 of the [Supplement for the TreeExplainer paper](https://static-content.springer.com/esm/art%3A10.1038%2Fs42256-019-0138-9/MediaObjects/42256_2019_138_MOESM1_ESM.pdf).

**Comment by brunofacca** [comment]
Thank you very much for the clarification, @HughChen. Using the Shapley value of the loss is an interesting idea, I will experiment with that.

**Comment by HughChen** [comment]
Sure thing - best of luck!

**Comment by eaishwa** [comment]
@HughChen By loss, do you mean shap values for log loss and not probability scores from the model?

**Comment by HughChen** [comment]
@eaishwa Yup!  Rather than explaining f(x) which outputs a probability prediction, you can explain L(f(x),y).  Here is an [example](https://github.com/slundberg/shap/blob/master/notebooks/tabular_examples/tree_based_models/Explaining%20the%20Loss%20of%20a%20Model.ipynb).

**Comment by cerlymarco** [comment]
Everyone looking for SHAP for feature selection (and hyperparameters tuning), I suggest [shap-hypetune](https://github.com/cerlymarco/shap-hypetune)

_A python package for simultaneous Hyperparameters Tuning and Features Selection for Gradient Boosting Models_

It supports feature selection with RFE or Boruta and parameter tuning with Grid or Random Search.
